lib/config/guiconfiguration.go

```diff
@@ -13,20 +13,20 @@ import (
 )
 
 type GUIConfiguration struct {
-	Enabled                   bool   `xml:"enabled,attr" json:"enabled" default:"true"`
-	RawAddress                string `xml:"address" json:"address" default:"127.0.0.1:8384"`
-	User                      string `xml:"user,omitempty" json:"user"`
-	Password                  string `xml:"password,omitempty" json:"password"`
-	RawUseTLS                 bool   `xml:"tls,attr" json:"useTLS"`
-	APIKey                    string `xml:"apikey,omitempty" json:"apiKey"`
-	InsecureAdminAccess       bool   `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
-	Theme                     string `xml:"theme" json:"theme" default:"default"`
-	Debugging                 bool   `xml:"debugging,attr" json:"debugging"`
-	InsecureSkipHostCheck     bool   `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
-	InsecureAllowFrameLoading bool   `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
+	Enabled                   bool     `xml:"enabled,attr" json:"enabled" default:"true"`
+	RawAddress                []string `xml:"address" json:"address" default:"127.0.0.1:8384"`
+	User                      string   `xml:"user,omitempty" json:"user"`
+	Password                  string   `xml:"password,omitempty" json:"password"`
+	RawUseTLS                 bool     `xml:"tls,attr" json:"useTLS"`
+	APIKey                    string   `xml:"apikey,omitempty" json:"apiKey"`
+	InsecureAdminAccess       bool     `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
+	Theme                     string   `xml:"theme" json:"theme" default:"default"`
+	Debugging                 bool     `xml:"debugging,attr" json:"debugging"`
+	InsecureSkipHostCheck     bool     `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
+	InsecureAllowFrameLoading bool     `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
```

Comment:
InsecureAllowFrameLoading, InsecureSkipHostCheck, InsecureAdminAccess, RawUseTLS should probably be per-address.

Reply:
Does it mean that we need to promote the `RawAddresses` field as a structure ?

Reply:
That, plus a migration, converting the old string to a new struct.
So `Deprecated_RawAddress`, `Deprecated_UseTLS`, etc etc, gets converted to:
```
GUIAddress{ 
  Addresss string
  UseTLS bool
  ...
}
```

I guess asking you to do that is unfair, as you are only interested in getting support for 2nd'ary addresses, but this is the way I'd imagine it works under ideal circumstances, so I am just throwing it out there.

Reply:
I will investigate that point.

Reply:
Here is a proposal for a new GUIConfiguration structuration. What's your opinion?

```
type GUIListener struct {
	RawAddress                string `xml:"address" json:"address" default:"127.0.0.1:8384"`
	RawUseTLS                 bool   `xml:"tls,attr" json:"useTLS"`
	InsecureAdminAccess       bool   `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
	InsecureSkipHostCheck     bool   `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
	InsecureAllowFrameLoading bool   `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
}

type GUIConfiguration struct {
	Listener  GUIListener `xml:"listener" json:"listener"`
	Enabled   bool        `xml:"enabled,attr" json:"enabled" default:"true"`
	User      string      `xml:"user,omitempty" json:"user"`
	Password  string      `xml:"password,omitempty" json:"password"`
	APIKey    string      `xml:"apikey,omitempty" json:"apiKey"`
	Theme     string      `xml:"theme" json:"theme" default:"default"`
	Debugging bool        `xml:"debugging,attr" json:"debugging"`

	// Deprecated. Old listener configuration style.
	RawAddresses              string `xml:"address" json:"address" default:"127.0.0.1:8384"`
	RawUseTLS                 bool   `xml:"tls,attr" json:"useTLS"`
	InsecureAdminAccess       bool   `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
	InsecureSkipHostCheck     bool   `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
	InsecureAllowFrameLoading bool   `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
}
```

Reply:
I'd also rename the old deprecated fields as Deprecated_, and add a config migration which converts old values to a single item to Listeners, and nulls them out. Also, you'd want to add omitEmpty to them, to remove them from showing up.

Reply:
Updated according to your observations:

```
type GUIListener struct {
	Address                   string `xml:"address" json:"address" default:"127.0.0.1:8384"`
	UseTLS                    bool   `xml:"tls,attr" json:"useTLS"`
	InsecureAdminAccess       bool   `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
	InsecureSkipHostCheck     bool   `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
	InsecureAllowFrameLoading bool   `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
}

type GUIConfiguration struct {
	Listeners []GUIListener `xml:"listener" json:"listeners"`
	Enabled   bool          `xml:"enabled,attr" json:"enabled" default:"true"`
	User      string        `xml:"user,omitempty" json:"user"`
	Password  string        `xml:"password,omitempty" json:"password"`
	APIKey    string        `xml:"apikey,omitempty" json:"apiKey"`
	Theme     string        `xml:"theme" json:"theme" default:"default"`
	Debugging bool          `xml:"debugging,attr" json:"debugging"`

	// Deprecated. Old listener configuration style.
	Deprecated_RawAddress                string `xml:"address,omitempty" json:"address"`
	Deprecated_RawUseTLS                 bool   `xml:"tls,attr,omitempty" json:"useTLS"`
	Deprecated_InsecureAdminAccess       bool   `xml:"insecureAdminAccess,omitempty" json:"insecureAdminAccess"`
	Deprecated_InsecureSkipHostCheck     bool   `xml:"insecureSkipHostcheck,omitempty" json:"insecureSkipHostcheck"`
	Deprecated_InsecureAllowFrameLoading bool   `xml:"insecureAllowFrameLoading,omitempty" json:"insecureAllowFrameLoading"`
}
```

Reply:
Maybe we should rename the `Addresses()` method to `Listeners()` for coherency.

Reply:
Also, do we consider to rename the environment variable `STGUIADDRESS` to `STGUIADDRESSES`? This will of course introduce a backward incompatibility. Or maybe we could support both forms.

Reply:
Sure. Yet you'll have to change the tags on the deprecated fields to prevent json serialization.

Reply:
And as I said, `default` value has no meaning in your case.

Reply:
Ok. I propose to handle both deprecated config migration and default listener injection in the `loadConfig()` function. Fine for you or do you see a better place?

Reply:
Migrations are handled in /lib/config/config.go or wrapper.go, you should just add a migration there.
Injecting a default listener should probably happen somewhere in prepare(), given Enabled is true.

Reply:
Fine.
My last blocking point concerns the JSON serialization part related to the deprecated fields. Where concretely should we prevent them to be serialized?

Reply:
Oh. The new configuration restructuration implies to rework the dump/get/set operations at the `stcli' command level. How should we cope with the new structuration at this level?

Reply:
The tool is half broken anyway, so we can many any needed breaking changes there.

Reply:
Work on the go code looks well. Now I definitely need some pointers for the Javascript part.
From my current comprehension, I have identified the following tasks:
- update the HTML layout for the Settings modal view. Maybe we could dynamically handle one   `<div>` per listener and gather all the listener properties (address, UseTLS, ...) inside each division. But I have no idea on how to handle the dynamic side;
- I suspect that the test related to `$scope.openNoAuth` should be performed on the right listener but I have no idea on how to identify the right one.
- and definitely some other impacts I have not yet identified ;-)

Reply:
Yeah, this will be an interesting problem as a lot of these settings will become conditional.
Best idea I have is to have a $scope.listeners [{useTls: true, addess:bar}, ..] then render a small template per listener, and have - button next to each entry to remove and a plus at the bottom to add.